util/benchmark/ec/p256_awslc_ossl/src/benchmark_ecdh.c

```diff
@@ -0,0 +1,183 @@
+    {
+
+        if (EVP_PKEY_keygen(kctx, &key_A) <= 0 || /* generate secret key A */
+            !(ctx = EVP_PKEY_CTX_new(key_A, NULL)) || /* derivation ctx from skeyA */
+            EVP_PKEY_derive_init(ctx) <= 0 || /* init derivation ctx */
+            EVP_PKEY_derive_set_peer(ctx, key_B) <= 0 || /* set peer pubkey in ctx */
+            EVP_PKEY_derive(ctx, NULL, &outlen) <= 0 || /* determine max length */
+            outlen == 0 ||  /* ensure outlen is a valid size */
+            outlen > MAX_ECDH_SIZE /* avoid buffer overflow */)
+        {
+            ecdh_checks = 0;
+            BIO_printf(bio_err, "ECDH secret length output failure.\n");
+            ERR_print_errors(bio_err);
+        }
+    }
+
+    if (1 == ecdh_checks)
+    {
+        /* Allocate buffer for shared secret and test shared secret */
+        if (NULL == (secret_a = OPENSSL_malloc(outlen)) ||
+            NULL == (secret_b = OPENSSL_malloc(outlen)))
+        {
+            ecdh_checks = 0;
+            BIO_printf(bio_err, "ECDH buffer allocation failure.\n");
+            ERR_print_errors(bio_err);
+        }
+    }
+
+    if (1 == ecdh_checks)
+    {
+        /* Test the key derivation by computing the shared secret on B's side */
+        if (!(test_ctx = EVP_PKEY_CTX_new(key_B, NULL)) || /* test ctx from skeyB */
+            !EVP_PKEY_derive_init(test_ctx) || /* init derivation test_ctx */
+            !EVP_PKEY_derive_set_peer(test_ctx, key_A) || /* set peer pubkey in test_ctx */
+            !EVP_PKEY_derive(test_ctx, NULL, &test_outlen) || /* determine max length */
+            !EVP_PKEY_derive(ctx, secret_a, &outlen) || /* compute a*B */
+            !EVP_PKEY_derive(test_ctx, secret_b, &test_outlen) || /* compute b*A */
+            test_outlen != outlen /* compare output length */ ||
+            CRYPTO_memcmp(secret_a, secret_b, outlen)) /* compare A's and B's shared secret bytes */
+        {
+            ecdh_checks = 0;
+            BIO_printf(bio_err, "ECDH shared secret test failure.\n");
+            ERR_print_errors(bio_err);
+        }
+    }
+
+    if (1 == ecdh_checks)
+    {
+        /* Test the key derivation after regenerating A's key using the same context */
+        if (EVP_PKEY_keygen(kctx, &key_A) <= 0 || /* generate secret key A again */
+            !EVP_PKEY_derive(ctx, secret_a, &outlen) || /* compute a*B */
+            !CRYPTO_memcmp(secret_a, secret_b, outlen) || /* ensure not same shared secret as before */
+            !EVP_PKEY_derive(test_ctx, secret_b, &test_outlen) || /* compute b*A */
+            CRYPTO_memcmp(secret_a, secret_b, outlen)) /* compare A's and B's shared secret bytes */
+        {
+            ecdh_checks = 0;
+            BIO_printf(bio_err, "ECDH shared secret with new A key test failure.\n");
+            ERR_print_errors(bio_err);
+        }
+    }
+
+    if (1 == ecdh_checks)
+    {
+        /* Warm up and instrument the function to calculate how many iterations it should run for */
```

Comment:
Not sure I understand the rational behind using this method. What is the benefit of running the benchmark a dynamic number of iterations? 
`report_results` computes the number of iterations over a specified period of time. If we have a limit on the number of iterations that can run, won't we risk hitting that limit before having run for the specified period of time?